src/miner.cpp

```diff
@@ -798,24 +934,27 @@ void GenerateDynamics(int nCPUThreads, int nGPUThreads, const CChainParams& chai
 
     boost::thread_group* cpuMinerThreads = GetCPUMinerThreads();
 
-    if (nCPUThreads < 0)
-        nCPUThreads = GetNumCores();
+    int nNumCores = GetNumCores();
+    LogPrintf("DynamicMiner -- CPU Cores: %u\n", nNumCores);
+
+    if (nCPUThreads < 0 || nCPUThreads > nNumCores)
+        nCPUThreads = nNumCores;
 
     // Start CPU threads
     while (cpuMinerThreads->size() < nCPUThreads) {
-        std::size_t nThread = cpuMinerThreads->size() -1;
+        std::size_t nThread = cpuMinerThreads->size();
         LogPrintf("Starting CPU Miner thread #%u\n", nThread);
         cpuMinerThreads->create_thread(boost::bind(&DynamicMiner, boost::cref(chainparams), boost::ref(connman), nThread, false));
     }
 
-    if (nGPUThreads < 0)
-        nGPUThreads = 1;
+    if (nGPUThreads < 0 || nGPUThreads > (int)devices)
+        nGPUThreads = devices;
 
     // Start GPU threads
     boost::thread_group* gpuMinerThreads = GetGPUMinerThreads();
     //for (std::size_t device = 0; device < devices; device++) {
         for (std::size_t i = 0; i < nGPUThreads || i < devices; i++) {
```

Comment:
This is probably why amount of threads wouldn't impact performance `nGPUThreads || i < devices`